src/particles/spacecharge/GatherAndPush.cpp

```diff
@@ -0,0 +1,107 @@
+
+                // get the device pointer-wrapper Array4 for 3D field access
+                auto const scf_arr_x = space_charge_field.at(lev).at("x")[pti].array();
+                auto const scf_arr_y = space_charge_field.at(lev).at("y")[pti].array();
+                auto const scf_arr_z = space_charge_field.at(lev).at("z")[pti].array();
+
+                // physical constants and reference quantities
+                amrex::ParticleReal const c0_SI = 2.99792458e8;  // TODO move out
+                amrex::ParticleReal const mc_SI = pc.GetRefParticle().mass * c0_SI;
+                amrex::ParticleReal const pz_ref_SI = pc.GetRefParticle().beta_gamma() * mc_SI;
+                amrex::ParticleReal const gamma = pc.GetRefParticle().gamma();
+                amrex::ParticleReal const inv_gamma2 = 1.0_prt / (gamma * gamma);
+
+                amrex::ParticleReal const dt = slice_ds / pc.GetRefParticle().beta() / c0_SI;
+
+                // preparing access to particle data: AoS
+                using PType = ImpactXParticleContainer::ParticleType;
+                auto const & aos = pti.GetArrayOfStructs();
+                PType const * const AMREX_RESTRICT aos_ptr = aos().dataPtr();
+
+                // preparing access to particle data: SoA of Reals
+                auto& soa_real = pti.GetStructOfArrays().GetRealData();
+                amrex::ParticleReal* const AMREX_RESTRICT part_px = soa_real[RealSoA::ux].dataPtr();
+                amrex::ParticleReal* const AMREX_RESTRICT part_py = soa_real[RealSoA::uy].dataPtr();
+                amrex::ParticleReal* const AMREX_RESTRICT part_pz = soa_real[RealSoA::pt].dataPtr(); // note: currently in z
+
+                // group together constants for the momentum push
+                amrex::ParticleReal const push_consts = dt * charge * inv_gamma2 / pz_ref_SI;
+
+                // gather to each particle and push momentum
+                amrex::ParallelFor(np, [=] AMREX_GPU_DEVICE (int i) {
+                    // access AoS data such as positions and cpu/id
+                    PType const & AMREX_RESTRICT p = aos_ptr[i];
+
+                    // access SoA Real data
+                    amrex::ParticleReal & AMREX_RESTRICT px = part_px[i];
+                    amrex::ParticleReal & AMREX_RESTRICT py = part_py[i];
+                    amrex::ParticleReal & AMREX_RESTRICT pz = part_pz[i];
+
+                    // force gather
+                    amrex::GpuArray<amrex::Real, 3> const field_interp =
+                        ablastr::particles::doGatherVectorFieldNodal (
+                            p.pos(0), p.pos(1), p.pos(2),
+                            scf_arr_x, scf_arr_y, scf_arr_z,
+                            invdr,
+                            prob_lo);
+
+                    // push momentum
+                    px += field_interp[0] * push_consts;
+                    py += field_interp[1] * push_consts;
+                    pz += field_interp[2] * push_consts;
```

Comment:
Status today using #244: it works! :tada:

We just need to find out why I had to locally hack in a `-` for the direction of the force ;-)

Reply:
Found it:
https://github.com/ECP-WarpX/impactx/blob/8e260c47e7d13f28da1477f213200efccee1755f/src/particles/ChargeDeposition.cpp#L69-L70

Reply:
Fixed via #251 and rebased